### PR TITLE
1206 test robin approximating dirichlet

### PR DESCRIPTION
**Pull request by IngridKJ**
## Proposed changes

Addresses issue #1206. Robin conditions are on the form flux + alpha * potential = G. As alpha goes to zero, the Robin conditions approaches Neumann. As alpha goes to infinity, the Robin conditions approaches Dirichlet. This PR includes a test to ensure that the latter happens. The test covers a scalar problem (mass and energy balance) as well as a vector problem (momentum balance). This PR also improves the setup in the approximate-Neumann-conditions-test which was already present. Special thanks to @IvarStefansson for great assistance!

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [x] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.
- [x] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [x] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: tests/models/test_boundary_condition.py [hunk 1]
```diff
@@ -113,57 +113,97 @@ def test_boundary_condition_mixin(t_end: int):
 """Here follows mixins related to testing of Robin limit cases, and eventually the test itself. """
 
 
-class BCValues:
-    """Set boundary values for momentum balance and mass and energy balance."""
+class BCValuesLeftSide:
```
Comment (IvarStefansson) [inline tests/models/test_boundary_condition.py:116]: The name is somewhat misleading.

File: tests/models/test_boundary_condition.py [hunk 2]
```diff
@@ -113,57 +113,97 @@ def test_boundary_condition_mixin(t_end: int):
...
+    * Momentum balance
+    * Mass and energy balance.
+
+    """
+
+    def not_dir_inds(self, g):
+        """Indices for the non-Dirichlet boundaries for test.
+
+        The Robin limit case test tests setups with Dirichlet on west/east boundaries,
+        and Robin approximating Dirichlet on the north/south boundaries. This method
```
Comment (IvarStefansson) [inline tests/models/test_boundary_condition.py:129]: Wouldn't it be right to say that we use Robin approximating either Dirichlet or Neumann?

File: tests/models/test_boundary_condition.py [hunk 3]
```diff
@@ -113,57 +113,97 @@ def test_boundary_condition_mixin(t_end: int):
...
+        """
+        bounds = self.domain_boundary_sides(g)
+        return bounds.west + bounds.east
 
     def bc_values_displacement(self, bg: pp.BoundaryGrid) -> np.ndarray:
-        """Assigns displacement values in the x-direction of the west boundary."""
-        values = np.ones((self.nd, bg.num_cells))
-        bounds = self.domain_boundary_sides(bg)
-        values[0][bounds.west] += np.ones(len(values[0][bounds.west]))
+        """Assigns displacement values in the x-direction of the west and east
```
Comment (IvarStefansson) [inline tests/models/test_boundary_condition.py:148]: Since the dir_inds method may be overridden, I suggest:
```suggestion
        """Assigns displacement values in the x-direction of the Dirichlet
```
If you agree, please roll out throughout the file.

File: tests/models/test_boundary_condition.py [hunk 4]
```diff
@@ -113,57 +113,97 @@ def test_boundary_condition_mixin(t_end: int):
...
-    """Set Dirichlet and Neumann for momentum balance and mass and energy balance.
+class BCRobin:
+    """Set Dirichlet and Robin for momentum balance and mass and energy balance.
+
+    Sets Dirichlet on the west and east boundaries, and Robin on the top and bottom
+    boundaries. The value of the Robin weight is determined from the parameter "alpha"
+    in the params dictionary. This class also sets Robin boundary values.
 
-    Sets Dirichlet on the west boundary, and Neumann on all other boundaries.
+    This class is common for all the model classes that enters into testing Robin limit
```
Comment (IvarStefansson) [inline tests/models/test_boundary_condition.py:176]: Is that true? Isn't it only used for the Robin version, not the reference model?

File: tests/models/test_boundary_condition.py [hunk 5]
```diff
@@ -173,134 +213,221 @@ def bc_type_fourier_flux(self, sd: pp.Grid) -> pp.BoundaryCondition:
...
         bounds = self.domain_boundary_sides(sd)
-        bc = pp.BoundaryConditionVectorial(sd, bounds.all_bf, "rob")
-        bc.is_rob[:, bounds.west] = False
-        bc.is_dir[:, bounds.west] = True
-
-        alpha = self.params["alpha"]
-
-        r_w = np.tile(np.eye(sd.dim), (1, sd.num_faces))
-        bc.robin_weight = np.reshape(r_w, (sd.dim, sd.dim, sd.num_faces), "F") * alpha
+        bc = pp.BoundaryConditionVectorial(sd, bounds.all_bf, "neu")
```
Comment (IvarStefansson) [inline tests/models/test_boundary_condition.py:226]: You could use the slightly more compact version I used for the scalar BCs

File: tests/models/test_boundary_condition.py (hunk 2 above)
Comment (IvarStefansson) [reply tests/models/test_boundary_condition.py:129]: And do you think it would be slightly easier to understand this method if we rename it to rob_inds?

File: tests/models/test_boundary_condition.py [hunk 6]
```diff
@@ -173,134 +213,221 @@ def bc_type_fourier_flux(self, sd: pp.Grid) -> pp.BoundaryCondition:
...
-            pressure = instance.pressure([sd]).value(instance.equation_system)
-            temperature = instance.temperature([sd]).value(instance.equation_system)
-            return {"temperature": temperature, "pressure": pressure}
+    params["alpha"] = alpha
+    instance = balance_class(params)
+    pp.run_time_dependent_model(instance, params)
+    sd = instance.mdg.subdomains(dim=2)[0]
 
-    return _run_model
+    try:
```
Comment (IvarStefansson) [inline tests/models/test_boundary_condition.py:384]: Isn't "isinstance(instance, MomentumBalance)" more elegant?

**Review by IvarStefansson — APPROVED** [review]
Good work! Please have a look at the small changes I suggested and merge when fixed.

File: tests/models/test_boundary_condition.py (hunk 4 above)
Comment (IngridKJ) [reply tests/models/test_boundary_condition.py:176]: You're correct, will fix!

File: tests/models/test_boundary_condition.py (hunk 2 above)
Comment (IngridKJ) [reply tests/models/test_boundary_condition.py]: I renamed to rob_inds, as I agree that can make things  more understandable. Might still be a bit "weird" that rob_inds is used in the Dirichlet reference class, but I think the upsides of changing the name are big enough to accept that slight "weirdness".
